### PR TITLE
Add chunked logit computation to SigLipLoss for memory efficiency

### PR DESCRIPTION
**Pull request by avocardio**
## Summary

Adds a `chunk_size` parameter to `SigLipLoss` that chunks the B×N logit matrix computation, reducing peak memory from O(B×N) to O(chunk_size×N). This addresses the memory scaling issues reported in #942 and #765.

**Usage:**
```python
loss_fn = SigLipLoss(chunk_size=512)  # chunks logits into 512-row blocks
```

Default `chunk_size=0` preserves the original behavior exactly — no changes to existing code paths.

All four distributed strategies (bidir, shift, reduce, gather) benefit automatically since they all call `_loss()` internally.

## Memory savings

At B=4096, D=768:
- Original: 64MB for logit matrix
- chunk_size=512: 8MB (8× less)

At B=32768 (SigLIP's optimal batch size):
- Original: 4GB for logit matrix
- chunk_size=512: 64MB (64× less)

## Changes

- Added `chunk_size` parameter to `SigLipLoss.__init__`
- Added `_chunked_loss()` method that chunks over image feature rows
- `_loss()` dispatches to `_chunked_loss()` when `chunk_size > 0`
- No new dependencies, pure PyTorch
- Zero changes to the default code path

## Related issues

- #942 — SigLip memory consumption increases with GPU count
- #765 — OOM with SigLIP at batch_size=1024

### CODE REVIEW

**Comment by avocardio** [comment]
## Memory savings across GPU architectures

Tested at D=768, fwd+bwd, f32. Peak memory savings from chunked logit computation:

| Batch | Reference | Chunked (cs=512) | Reduction |
|-------|-----------|------------------|-----------|
| 16,384 | 4.2 GB | 288 MB | **14.7×** |
| 32,768 | 16.7 GB | 576 MB | **28.9×** |
| 65,536 | 66.1 GB | 1.15 GB | **57×** |

**Speed:** this is a pure-PyTorch loop, so the per-chunk launch overhead means wall time is comparable to or slightly slower than the unchunked path; expect no speedup. The win is enabling batch sizes that previously OOM'd. For users who also want a speedup, we maintain a Triton-kernel variant at https://github.com/avocardio/siglip-kernel that fuses the matmul and sigmoid BCE in a single pass.

**Comment by rwightman** [comment]
@avocardio so I took a look at this, it doesn't appear to work?

**Comment by avocardio** [comment]
@rwightman Sorry, fixed in 357bd97 — added a parity test.

**Comment by rwightman** [comment]
@avocardio k, that's working now... but, FWIW when I was looking at this I ran it by code and it suggested a further level of optimization that reduces allocations and should be equivalent down to floating point rounding


```
  def _chunked_loss(self, image_features, text_features, logit_scale, logit_bias=None, negative_only=False):
      B = image_features.shape[0]
      N = text_features.shape[0]
      chunk_size = min(self.chunk_size, B)

      total_loss = torch.zeros((), device=image_features.device, dtype=torch.float32)

      for i in range(0, B, chunk_size):
          end_i = min(i + chunk_size, B)
          img_chunk = image_features[i:end_i]
          logits = self.get_logits(img_chunk, text_features, logit_scale, logit_bias)

          # Treat every pair as negative:
          # -logsigmoid(-logits) == softplus(logits)
          chunk_loss = F.softplus(logits).sum()

          if not negative_only:
              # Replace local positives with positive-pair loss:
              # softplus(-x) - softplus(x) == -x
              num_pos = max(0, min(end_i, N) - i)
              if num_pos > 0:
                  rows = torch.arange(num_pos, device=logits.device)
                  pos_logits = logits[rows, i + rows]
                  chunk_loss = chunk_loss - pos_logits.sum()

          total_loss = total_loss + chunk_loss

      return total_loss / B
```

**Comment by rwightman** [comment]
Reasoning provided, 

The current loss is:

  -F.logsigmoid(labels * logits).sum()

  For a negative pair, labels = -1, so:

  -F.logsigmoid(-logit)

  That is exactly:

  F.softplus(logit)

  For a positive pair, labels = +1, so:

  -F.logsigmoid(logit)

  That is exactly:

  F.softplus(-logit)

  Since almost every entry is negative, we can compute the whole chunk as if every pair were negative:

  loss = F.softplus(logits).sum()

  Then fix only the diagonal positives. For each positive logit x, we currently counted:

  softplus(x)

  but we need:

  softplus(-x)

  So add the correction:

  loss += softplus(-x) - softplus(x)

  Using the identity:

  softplus(-x) - softplus(x) == -x

  that correction can be simplified to:

  loss -= x

  So no label matrix is needed. The labels are just encoding “all negative except diagonal positive,” and that structure can be represented directly with the default negative loss plus a diagonal correction.

**Comment by avocardio** [comment]
Nice — applied in 759f36f. Parity test still passes.

**Comment by rwightman** [comment]
@avocardio thanks, merging now
